File: 01_hello_usdz/main.swift
// MIT License
// 
// Copyright (C) 2018-2024, Tellusim Technologies Inc. https://tellusim.com/
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Tellusim

/*
 */
func create_texture(_ device: Device, _ material : MeshMaterial, _ type: String) -> Texture {
	
	// find material parameter
	let index = material.findParameter(type)
	if index == Maxu32 || !material.hasParameterFlag(index, MeshMaterial.Flags.Blob) { return Texture.null() }
	
	// load image
	let image = Image()
	let blob = material.getParameterBlob(index)
	if !image.load(blob) { return Texture.null() }
	
	// create texture
	return device.createTexture(image, Texture.Flags.Mipmaps)
}

/*
 */
func main() -> Int32 {
	
	// create app
	let app = App()
	if !app.create(Platform.Any, UInt32(App.Values.Version.rawValue)) { return 1 }
	
	// create window
	let window = Window(app.getPlatform(), app.getDevice())
	if !window || !window.setSize(app.getWidth(), app.getHeight()) { return 1 }
	if !window.create("01 Hello USDZ") || !window.setHidden(false) { return 1 }
	window.setKeyboardPressedCallback(Window.KeyboardPressedFunction({ (key: UInt32, code: UInt32) in
		if key == Window.Key.Esc.rawValue { window.stop() }
	}))
	
	// create device
	let device = Device(window)
	if !device { return 1 }
	
	// create pipeline
	let pipeline = device.createPipeline()
	pipeline.setSamplerMask(0, Shader.Mask.Fragment)
	pipeline.setTextureMasks(0, 3, Shader.Mask.Fragment)
	pipeline.setUniformMasks(0, 2, Shader.Mask.Vertex)
	pipeline.addAttribute(Pipeline.Attribute.Position, Format.RGBf32, 0, 0, 80)
	pipeline.addAttribute(Pipeline.Attribute.Normal, Format.RGBf32, 0, 12, 80)
	pipeline.addAttribute(Pipeline.Attribute.Tangent, Format.RGBAf32, 0, 24, 80)
	pipeline.addAttribute(Pipeline.Attribute.TexCoord, Format.RGf32, 0, 40, 80)
	pipeline.addAttribute(Pipeline.Attribute.Weights, Format.RGBAf32, 0, 48, 80)
	pipeline.addAttribute(Pipeline.Attribute.Joints, Format.RGBAu32, 0, 64, 80)
	pipeline.setColorFormat(window.getColorFormat())
	pipeline.setDepthFormat(window.getDepthFormat())
	pipeline.setDepthFunc(Pipeline.DepthFunc.Less)
	if !pipeline.loadShaderGLSL(Shader.Kind.Vertex, "main.shader", "VERTEX_SHADER=1") { return 1 }
	if !pipeline.loadShaderGLSL(Shader.Kind.Fragment, "main.shader", "FRAGMENT_SHADER=1") { return 1 }
	if !pipeline.create() { return 1 }
	
	// load mesh
	let mesh = Mesh()
	if !mesh.load("model.usdz") { return 1 }
	if mesh.getNumGeometries() == 0 { return 1 }
	if mesh.getNumAnimations() == 0 { return 1 }
	mesh.setBasis(Mesh.Basis.ZUpRight)
	mesh.createTangents()
	
	// create model
	let model = MeshModel()
	if !model.create(device, pipeline, mesh) { return 1 }
	
	// create textures
	var normal_textures = [Texture]()
	var diffuse_textures = [Texture]()
	var roughness_textures = [Texture]()
	for geometry in mesh.getGeometries() {
		for material in geometry.getMaterials() {
			normal_textures.append(create_texture(device, material, MeshMaterial.TypeNormal))
			diffuse_textures.append(create_texture(device, material, MeshMaterial.TypeDiffuse))
			roughness_textures.append(create_texture(device, material, MeshMaterial.TypeRoughness))
		}
	}
	
	// create sampler
	let sampler = device.createSampler(Sampler.Filter.Trilinear, Sampler.WrapMode.Repeat)
	if !sampler { return 1 }
	
	// create target
	let target = device.createTarget(window)
	target.setClearColor(Color.gray)
	
	// main loop
	window.run(Window.MainLoopFunction({ () -> Bool in
		
		// update window
		Window.update()
		
		// render window
		if !window.render() { return false }
		
		// window target
		if target.begin() {
			
			// create command list
			let command = device.createCommand(target)
			
			// set pipeline
			command.setPipeline(pipeline)
			
			// set sampler
			command.setSampler(0, sampler)
			
			// set model buffers
			model.setBuffers(command)
			
			// set common parameters
			struct CommonParameters {
				var projection: Matrix4x4f
				var modelview: Matrix4x4f
				var camera: Vector4f
			}
			let camera = Vector4f(0.0, -180.0, 180.0, 0.0)
			var common_parameters = CommonParameters(
				projection: Matrix4x4f.perspective(60.0, Float32(window.getWidth()) / Float32(window.getHeight()), 0.1, 1000.0),
				modelview: Matrix4x4f.lookAt(camera.xyz, Vector3f(0.0, 0.0, 80.0), Vector3f(0.0, 0.0, 1.0)),
				camera: camera
			)
			if target.isFlipped() { common_parameters.projection = Matrix4x4f.scale(1.0, -1.0, 1.0) * common_parameters.projection }
			command.setUniform(0, &common_parameters)
			
			// mesh animation
			let time = Time.seconds()
			let animation = mesh.getAnimation(0)
			animation.setTime(time, Matrix4x3d.rotateZ(sin(time) * 30.0))
			
			// draw geometries
			var texture_index = Int(0)
			var joint_parameters = [Vector4f](repeating: Vector4f.zero, count: 192)
			for geometry in mesh.getGeometries() {
				
				// joint transforms
				for i in 0 ..< geometry.getNumJoints() {
					let joint = geometry.getJoint(i)
					let transform = Matrix4x3f(animation.getGlobalTransform(joint)) * joint.getITransform() * geometry.getTransform()
					joint_parameters[Int(i) * 3 + 0] = transform.row_0
					joint_parameters[Int(i) * 3 + 1] = transform.row_1
					joint_parameters[Int(i) * 3 + 2] = transform.row_2
				}
				command.setUniform(1, &joint_parameters)
				
				// draw materials
				for material in geometry.getMaterials() {
					command.setTexture(0, normal_textures[texture_index])
					command.setTexture(1, diffuse_textures[texture_index])
					command.setTexture(2, roughness_textures[texture_index])
					model.draw(command, geometry.getIndex(), material.getIndex())
					texture_index += 1
				}
			}
			
			target.end()
		}
		
		// present window
		if !window.present() { return false }
		
		// check errors
		if !device.check() { return false }
		
		return true
	}))
	
	// finish context
	window.finish()
	
	return 0
}

/*
 */
print(main())

File: 01_hello_usdz/main.cs
// MIT License
// 
// Copyright (C) 2018-2024, Tellusim Technologies Inc. https://tellusim.com/
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

using System;
using Tellusim;
using System.Collections.Generic;
using System.Runtime.InteropServices;

/*
 */
class USDZ {
	
	/*
	 */
	[StructLayout(LayoutKind.Sequential)]
	public struct CommonParameters {
		public Matrix4x4f projection;
		public Matrix4x4f modelview;
		public Vector4f camera;
	}
	
	/*
	 */
	static Texture create_texture(Device device, MeshMaterial material, string type) {
		
		// find material parameter
		uint index = material.findParameter(type);
		if(index == Base.Maxu32 || !material.hasParameterFlag(index, MeshMaterial.Flags.Blob)) return null;
		
		// load image
		Image image = new Image();
		Blob blob = material.getParameterBlob(index);
		if(!image.load(blob)) return null;
		
		// create texture
		return device.createTexture(image, Texture.Flags.Mipmaps);
	}
	
	/*
	 */
	[STAThread]
	static void Main(string[] args) {
		
		// create app
		App app = new App(args);
		if(!app.create(Platform.Any)) return;
		
		// create window
		Window window = new Window(app.getPlatform(), app.getDevice());
		if(!window || !window.setSize(app.getWidth(), app.getHeight())) return;
		if(!window.create("01 Hello USDZ CSharp") || !window.setHidden(false)) return;
		window.setKeyboardPressedCallback((uint key, uint code, IntPtr data) => { if(key == (uint)Window.Key.Esc) window.stop(); });
		
		// create device
		Device device = new Device(window);
		if(!device) return;
		
		// create pipeline
		Pipeline pipeline = device.createPipeline();
		pipeline.setSamplerMask(0, Shader.Mask.Fragment);
		pipeline.setTextureMasks(0, 3, Shader.Mask.Fragment);
		pipeline.setUniformMasks(0, 2, Shader.Mask.Vertex);
		pipeline.addAttribute(Pipeline.Attribute.Position, Format.RGBf32, 0, 0, 80);
		pipeline.addAttribute(Pipeline.Attribute.Normal, Format.RGBf32, 0, 12, 80);
		pipeline.addAttribute(Pipeline.Attribute.Tangent, Format.RGBAf32, 0, 24, 80);
		pipeline.addAttribute(Pipeline.Attribute.TexCoord, Format.RGf32, 0, 40, 80);
		pipeline.addAttribute(Pipeline.Attribute.Weights, Format.RGBAf32, 0, 48, 80);
		pipeline.addAttribute(Pipeline.Attribute.Joints, Format.RGBAu32, 0, 64, 80);
		pipeline.setColorFormat(window.getColorFormat());
		pipeline.setDepthFormat(window.getDepthFormat());
		pipeline.setDepthFunc(Pipeline.DepthFunc.Less);
		if(!pipeline.loadShaderGLSL(Shader.Type.Vertex, "main.shader", "VERTEX_SHADER=1")) return;
		if(!pipeline.loadShaderGLSL(Shader.Type.Fragment, "main.shader", "FRAGMENT_SHADER=1")) return;
		if(!pipeline.create()) return;
		
		// load mesh
		Mesh mesh = new Mesh();
		if(!mesh.load("model.usdz")) return;
		if(mesh.getNumGeometries() == 0) return;
		if(mesh.getNumAnimations() == 0) return;
		mesh.setBasis(Mesh.Basis.ZUpRight);
		mesh.createTangents();
		
		// create model
		MeshModel model = new MeshModel();
		if(!model.create(device, pipeline, mesh)) return;
		
		// create textures
		List<Texture> normal_textures = new List<Texture>();
		List<Texture> diffuse_textures = new List<Texture>();
		List<Texture> roughness_textures = new List<Texture>();
		for(uint i = 0; i < mesh.getNumGeometries(); i++) {
			MeshGeometry geometry = mesh.getGeometry(i);
			for(uint j = 0; j < geometry.getNumMaterials(); j++) {
				MeshMaterial material = geometry.getMaterial(j);
				normal_textures.Add(create_texture(device, material, "normal"));
				diffuse_textures.Add(create_texture(device, material, "diffuse"));
				roughness_textures.Add(create_texture(device, material, "roughness"));
			}
		}
		
		// create sampler
		Sampler sampler = device.createSampler(Sampler.Filter.Trilinear, Sampler.WrapMode.Repeat);
		if(!sampler) return;
		
		// create target
		Target target = device.createTarget(window);
		target.setClearColor(Color.gray);
		
		// main loop
		window.run((IntPtr data) => {
			
			// update window
			Window.update();
			
			// render window
			if(!window.render()) return false;
			
			// window target
			target.begin();
			{
				// create command list
				Command command = device.createCommand(target);
				
				// set pipeline
				command.setPipeline(pipeline);
				
				// set sampler
				command.setSampler(0, sampler);
				
				// set model buffers
				model.setBuffers(command);
				
				// common parameters
				CommonParameters common_parameters = new CommonParameters();
				common_parameters.camera = new Vector4f(0.0f, -180.0f, 180.0f, 0.0f);
				common_parameters.projection = Matrix4x4f.perspective(60.0f, (float)window.getWidth() / (float)window.getHeight(), 0.1f, 1000.0f);
				common_parameters.modelview = Matrix4x4f.lookAt(new Vector3f(common_parameters.camera), new Vector3f(0.0f, 0.0f, 80.0f), new Vector3f(0.0f, 0.0f, 1.0f));
				if(target.isFlipped()) common_parameters.projection = Matrix4x4f.scale(1.0f, -1.0f, 1.0f) * common_parameters.projection;
				command.setUniform(0, common_parameters);
				
				// mesh animation
				double time = Time.seconds();
				MeshAnimation animation = mesh.getAnimation(0);
				animation.setTime(time, Matrix4x3d.rotateZ(Math.Sin(time) * 30.0));
				
				// draw geometries
				int texture_index = 0;
				Matrix4x3f[] joint_parameters = new Matrix4x3f[64];
				for(uint j = 0; j < mesh.getNumGeometries(); j++) {
					MeshGeometry geometry = mesh.getGeometry(j);
					
					// joint transforms
					for(uint i = 0; i < geometry.getNumJoints(); i++) {
						MeshJoint joint = geometry.getJoint(i);
						joint_parameters[i] = new Matrix4x3f(animation.getGlobalTransform(joint)) * joint.getITransform() * geometry.getTransform();
					}
					command.setUniform(1, joint_parameters);
					
					// draw materials
					for(uint i = 0; i < geometry.getNumMaterials(); i++) {
						MeshMaterial material = geometry.getMaterial(i);
						command.setTexture(0, normal_textures[texture_index]);
						command.setTexture(1, diffuse_textures[texture_index]);
						command.setTexture(2, roughness_textures[texture_index]);
						model.draw(command, geometry.getIndex(), material.getIndex());
						texture_index++;
					}
				}
				
				command.destroyPtr();
			}
			target.end();
			
			// present window
			if(!window.present()) return false;
			
			// check device
			if(!device.check()) return false;
			
			// memory
			GC.Collect();
			GC.WaitForPendingFinalizers();
			
			return true;
		});
		
		// finish context
		window.finish();
		
		// keep window alive
		window.unacquirePtr();
	}
}
